C99 support

nwc10 opened the pull request:
Tested on *nix.

Needs testing on Win32 and VMS. Also needs the Win32 documentation changes reviewed by someone familiar with Win32

nwc10 committed:
Teach Configure and cflags.SH about C99

Probe to see whether we need -std=gnu99 or -std=c99 to get C99 code to
compile. In cflags.SH, remove code that added gcc warning flags that were
compatible with C89 but are not compatible with C99.

nwc10 committed:
Update perlhack with details of the C99 features we can now use

Also update INSTALL to mention C99.

nwc10 committed:
Update README.win32 to the new C99/MSVC 12.0 requirements

* Remove specific notes related to Microsoft Visual C++ 2005 Express Edition
  and earlier
* Visual Studio 2013 is the only version that still has Express edition (in
  addition to Community), so simplify the text by only mentioning Community
* Remove the reference to IA64, and mention x86_64 along with x86
* In the context of x86_64, clarify that mingw's binaries can run on both
* Update the URL for mingw

nwc10 committed:
Introduce a "declaration after statement" into inline.h

The core now permits some C99 code, so extensions need to ensure that any
C compiler flags they add or change permit C99 code.

This code is inlined by XS extensions in the core and on CPAN. The intent
is that this commit should show up for any failure bisection, making it
obvious what the cause is, and what the fix needs to be.

nwc10 committed:
Note how best to confirm that other C99 features can be relied upon

If C99 functionality can't be emulated on platforms which don't provide it,
then one really needs to test whether MSVC and the VMS compiler support it.

steve-m-hay commented on README.win32:
```diff
@@ -63,16 +63,25 @@ that are also supported by perl's makefile.
-and are the same compilers that ship with "Visual C++ .NET 2003 Professional"
-or "Visual C++ 2005-2019 Professional" respectively.
+and are the same compilers that ship with
+or "Visual C++ 2013-2019 Professional".
```
The "or "is no longer wanted here.

nwc10 replied:
Well spotted. Fixed. Thanks.

steve-m-hay commented on README.win32:
```diff
@@ -63,16 +63,25 @@ that are also supported by perl's makefile.
       Microsoft Platform SDK	Nov 2001 (64-bit compiler and tools)
       MinGW64 compiler (gcc version 4.4.3 or later)
 
+but these have not been tested recently.
```
MinGW-w64 _is_ still being tested.

nwc10 replied:
The full context was:

```
This port can also be built on IA64/AMD64 using:

      Microsoft Platform SDK    Nov 2001 (64-bit compiler and tools)
      MinGW64 compiler (gcc version 4.4.3 or later)
```

and that text (as-was) is already ambiguous, isn't it? In that **IA64** is Itanium (and long dead for Windows) whereas **AMD64** is the new normal. So is it fair to say that we should just take out the whole thing, and change the earlier text:

```
This set of instructions is meant to describe a so-called "native"
port of Perl to the Windows platform.  This includes both 32-bit and
64-bit Windows operating systems.  The resulting Perl requires no
additional software to run (other than what came with your operating
system).  Currently, this port is capable of using one of the
following compilers on the Intel x86 architecture:

      Microsoft Visual C++    version 12.0 or later
      Intel C++ Compiler      (experimental)
      Gcc by mingw.org        gcc version 3.4.5-5.3.0
      Gcc by mingw-w64.org    gcc version 4.4.3 or later

Note that the last two of these are actually competing projects both
delivering complete gcc toolchain for MS Windows:
```

to reflect both on x86 and x86_64?


Or should instead keep that as-is and change the first text to:

```
This port can also be built on AMD64 using:

      MinGW64 compiler (gcc version 4.4.3 or later)
```

just with nicer formatting?

xenu replied:
The former sounds good to me. There's nothing special about AMD64.

nwc10 replied:
I removed IA64 entirely, and moved the mention of x86_64 to the earlier paragraph shown.

And then realised that it makes mingw a bit ambiguous without further clarification - unlike the others mingw is just a 32 bit toolchain, but what it build can **run** on 64 bit Windows. So I tweaked the text further as part of 0788bfbebf59728

I assume that if we like this (or we like it with further edits) then I will merge that commit/all commits changing README.win32 into one commit will all the changes, and a unified commit message.

nwc10 replied:
@steve-m-hay I think that the text now in README.win32 fixes the problem you spotted, so I'm marking this "conversation" as resolved.

If I got this wrong, please accept my apologies and re-open this "conversation".

xenu commented on README.win32:
```diff
@@ -63,16 +63,25 @@ that are also supported by perl's makefile.
 
 The Microsoft Visual C++ compilers are also now being given away free. They are
-available as "Visual C++ Toolkit 2003" or "Visual C++ 2005-2019 Express [or
+available as "Visual C++ 2013-2019 Express [or
```
Visual Studio 2013 is the only version that still has Express edition (in addition to Community), so mentioning both seems overly prudent. I think "Express/Community" can be now simplified to "Community" everywhere in this document.

nwc10 replied:
Done as part of commit 0788bfbebf59728 - see previous comment about merge/rebase.

smpetershm reviewed (commented):
This makes me twitch after adding that warning so many years ago.

steve-m-hay reviewed (changes requested):
win32/GNUmakefile and win32/Makefile also have large chunks that can be removed.
Likewise many .c and .h files (not just in win32/). VC12 has _MSC_VER 1800...

nwc10 commented:
> win32/GNUmakefile and win32/Makefile also have large chunks that can be removed. Likewise many .c and .h files (not just in win32/). VC12 has _MSC_VER 1800...

Oh my. That's a good thing, isn't it?

However, I'm not competent to do that. Is it better if we wait to do that in a follow up branch/PR?

smpetershm commented:
@nwc10 Just watching the `-Wdeclaration-after-statement` that I added because GCC 3.0 stopped enforcing it and my OpenBSD 368 was stuck on 2.95.  Back in the old days when `-Wall` was really scary.

nwc10 committed:
Devel::PPPort shouldn't unconditionally -Wdeclaration-after-statement

Add the flag for gcc for perl v5.34.4 and earlier.

jkeenan commented:
On 10/6/21 2:27 PM, Steve Peters wrote:
>     Almost as good news: I tried this branch on FreeBSD-12 using |gcc9|
>     (the default version of |gcc| on this platform). There were 3 new
>     instances of the |-Wdeclaration-after-statement| warnings category.
>     I suspect these warnings can be remediated fairly easily.
> 
> C99 allows for mixed declarations and statements so checking for 
> |-Wdeclaration-after-statement| won't be necessary for this change or after.
> 

How then do we quiet this warning, at least with respect to these 3 
instances?

nwc10 committed:
perldelta for C99 support

nwc10 commented:
> Should we now document that configuring with `-Accflags="-Werror=declaration-after-statement"` is now inappropriate? (I want to have the correct info before I contact this smoke-tester.)

I guess that we should wait until reviewers have said that this PR is good to merge.

Also, I'm not sure **where** to document that this is no longer a good smoke test configuration **for blead** - does it belong in core, or is there somewhere in the smoker ecoystem?

nwc10 commented:
> [Grepping CPAN](https://grep.metacpan.org/search?size=20&q=Werror%3Ddeclaration-after-statement&qd=&qft=) suggests that there may be other scalps to come along in Sereal-Decoder, -Encoder, -Merger, -Path, -Splitter and YAML-Safe.

Thanks. I was aware of Sereal, hence Sereal/Sereal#265 (partly because in turn Yves is aware of this C99 work and the implications).

I wasn't aware of YAML-Safe, but I know that Cpanel::JSON::XS will need a tweak.

I was assuming that PRs (or similar) on these modules should wait until we're ready to merge, and was hoping that there would be a week or two between that point and when 5.35.5 ships, such that we have time for everyone to be ready.

tonycoz commented on pod/perlhacktips.pod:
```diff
@@ -111,6 +112,176 @@ L<perlguts>.
+        char message[];
+    };
+
+However, the source code already uses the "unwarranted chumminess with the
+compiler" hack in many places:
+
+    struct greeting {
+        unsigned int len;
+        char message[1];
+    };
```
Accessing beyond message[0] is technically undefined behaviour, so we avoid that.  I don't know of anywhere it's a practical issue though.

tonycoz commented on pod/perlhacktips.pod:
```diff
@@ -111,6 +112,176 @@ L<perlguts>.
+
+The VMS toolchain only added support for C<PRIdN> etc very recently, meaning
+that there are live supported installions without this, or formats such as
+C<%zu>
```
Perl's own formatters support %zu, so this might be confusing.

nwc10 replied:
Yes, well spotted. `Perl_sv_vcatpvfn_flags` takes `%`. It's only the C library `*printf` functions where this matters, and we "outlaw" most of those except `sprintf`. I was typing "maybe I should scrub this" but I think it better to clarify `sprintf` vs sv_catpvf etc, as we do use `sprintf` in our C code, and it will be a trap.

nwc10 replied:
I tried to clarify this in commit 41a65e5d9d44cfe10f3 - is this better?

tonycoz replied:
That looks good.

nwc10 committed:
Test declarations after statement in Configure's C99 probe code

Also ensure that the relevant failure error message is output even with
Configure's -s flag, as we shouldn't stay silent for a message that causes
Configure default to aborting.

With these changes, Configure will fail the C99 probe test if passed
-Accflags="-Werror=declaration-after-statement" or
-Accflags="-Werror=long-long"

nwc10 committed:
cflags.sh should not add -std=c99 if -std=gnu99 is already set

Having both looks like it ought to cause the build to fail, as with -std=c99
second it will override any previous -std=gnu99, and some platforms need the
former to both compile C99 code and have correct includes during Configure
probing.

However testing on those platforms without this change doesn't show any
errors. So this change doesn't seem to strictly be necessary. But it does
prevent our generated compiler command lines having contradictory flags,
and hence looking daft.

nwc10 committed:
Add a section to INSTALL describing that we now rely on some C99

And the implications for XS authors writing code to work on both current and
earlier perl installations.
